Add pull request template (#1)
cyclus/.github: .github/pull_request_template.md

<!-- If this PR adds a CEP, do not replicate information in the document. Give a brief summary and link to related issues and PRs. -->

# Summary of Changes
<!--- In one or more sentences, describe the PR you are submitting. -->

# Related CEPs and Issues

> <!--- Reference any related CEPs or issues and describe how this PR contributes to or addresses the problem. -->

Reviewer: Should we add more here encouraging (requiring?) folks ensure that at least one issue exists already? And then the template can include
```
Fixes # 
```

as a prompt to reference that/those issues.

Reviewer 2: I like the Fixes prompt. We could also update the comment to specify that an issue must exist before a PR can be made. I think MOOSE does that, and has all conversation in the issue (someone that actually interacts with that community can fact check that). I think this fits in with the "should we have a contributing guidelines document at the org level" that we have in #2. 

Related -- Do we want to open the discussions feature for this org? With the decline (or at least I feel like its declining) of the google email group, this could be a more accessible way for folks to ask questions about Cyclus. I've used the feature to ask questions in other repos when it's not really an issue. If we open this feature, then we can have a discussion about org-level contributing guidelines 😄

Author: I like the suggestion of adding the `Fixes #` prompt!

```suggestion
<!--- Reference the issue that this PR closes, any related CEPs, and describe how this PR contributes to or addresses the problem. -->
Fixes #
```

I haven't used the discussion feature before, but it might reduce the number of platforms we engage with people on?

# Associated Developers
<!--- Please mention any developers who should be alerted of this PR. -->

# Design Notes

> <!--- Describe any design decisions or trade-offs you made. -->

Reviewer 2: I'm not sure how I feel about this section. I feel like this information should be in the summary of changes. Another section that I have seen for templates is type of change (e.g., bug fix, new feature, breaking change, etc.). Is that something we want to have instead of Design Notes?

Reviewer: > Another section that I have seen for templates is type of change (e.g., bug fix, new feature, breaking change, etc.). Is that something we want to have instead of Design Notes?

Ideally (and we never quite live up to that ideal) each PR will be responding to an issue and the issue will indicate what type it is...

Author: I can understand wanting to include the information in the summary section. My thinking here is: if we don't explicitly ask for it, then the summary might just be "I did this thing." Whereas, if we ask for them to highlight design decisions or trade-offs, the author at least has to scroll past it before they make an issue. 

Maybe a clearer structure here would be to make this a subsection of the Summary of Changes?

Reviewer 2: I like the idea of making this a subsection of Summary of Changes.

> # Summary of Changes
> <!--- In one or more sentences, describe the PR you are submitting. -->
> 
> 
> # Related CEPs and Issues
> <!--- Reference any related CEPs or issues and describe how this PR contributes to or addresses the problem. -->
> 
> 
> # Associated Developers
> <!--- Please mention any developers who should be alerted of this PR. -->
> 
> 
> # Design Notes
> <!--- Describe any design decisions or trade-offs you made. -->

Author: What do you think of this @abachma2?
```suggestion
# Summary of Changes
<!--- In one or more sentences, describe the PR you are submitting. -->

## Design Notes
<!--- Describe any design decisions or trade-offs you made. -->

# Related CEPs and Issues
<!--- Reference any related CEPs or issues and describe how this PR contributes to or addresses the problem. -->

# Associated Developers
<!--- Please mention any developers who should be alerted of this PR. -->

```

Reviewer 2: I like this format.

# Testing and Validation
<!--- Describe how this change was tested. -->

# Checklist

> - [ ] Read the [Contributing to Cyclus](https://fuelcycle.org/kernel/contributing_to_cyclus.html) guide.
> - [ ] Compile and run locally.
> - [ ] Add or update tests.
> - [ ] Document if needed.
> - [ ] Follow style guidelines.
> - [ ] Update the changelog.
> - [ ] Address all review comments.

Reviewer: [pedantic] I don't think we need periods at the end of these
```suggestion
- [ ] Read the [Contributing to Cyclus](https://fuelcycle.org/kernel/contributing_to_cyclus.html) guide
- [ ] Compile and run locally
- [ ] Add or update tests
- [ ] Document if needed
- [ ] Follow style guidelines
- [ ] Update the changelog
- [ ] Address all review comments
```

Author: I understand removing periods for a less formal vibe! In undergrad we were told to end list items with punctuation, and I've done that ever since. Both APA and MLA guides agree that complete imperatives should have a period, but the newer versions leave options incomplete items.

Reviewer: I'll accept that formally, these probably should have punctuation because they are imperitives and not lists of items (for which commas would be better), but it did look a little abrupt... (and I think that even though I'm Gen-X 🤣 )

Author: Great point!

> <!-- If this PR adds a CEP, do not replicate information in the document. Give a brief summary and link to related issues and PRs. -->

Reviewer 3: Should this be a comment in the summary section?

Reviewer 4: I agree this should be moved into the comment under `Summary of changes`.

Reviewer 4: What do you think of this? Also, the second clause is covered by lines #16-18 here. 

```suggestion
<!-- If this PR adds a CEP, do not repeat all of the information in the CEP document in the summary. Instead, provide a short description of the CEP's purpose.  -->
```

# Summary of Changes
<!--- In one or more sentences, describe the PR you are submitting. -->

## Design Notes
<!--- Describe any design decisions or trade-offs you made. -->

Is this change expected to break any of the following?
- [ ] API for Cyclus modules
- [ ] Input files

> - [ ] Output tables for post processing tools

Reviewer 3: This may be a "my monkey brain" thing, but perhaps making it so that you check the box when it DOESN'T break them? I think that will make it display "correctly" as like 10/10 in the PR summary, too.

Author: That's a great point! I'll do that now

# Related CEPs and Issues
<!--- Reference the issue that this PR closes, any related CEPs, and describe how this PR contributes to or addresses the problem. -->

> Fixes #

Reviewer 4: ```suggestion
```

Unless we explicitly require an issue to be closed, I think this isn't necessary in the template.

Reviewer 3: I agree

Reviewer 3: Actually, on second thought, we DO probably want to enforce that people make PRs close issues...

Author: This line came from a comment on the original PR from @gonuke, and I do agree with him that it is nice to encourage people to do this. The templates are a suggestion after all, the only things they need to include are from CEP 6!

Reviewer 4: I see the comments, which were resolved. I'm fine disregarding this suggestion!

# Associated Developers
<!--- Please mention any developers who should be alerted of this PR. -->

# Testing and Validation
<!--- Describe how this change was tested. -->

- [ ] Read the [Contributing to Cyclus](https://fuelcycle.org/kernel/contributing_to_cyclus.html) guide
- [ ] Compile and run locally
- [ ] Add or update tests
- [ ] Document if needed
- [ ] Follow style guidelines
- [ ] Update the changelog

> - [ ] Address all review comments

Reviewer 3: One thing I've been adding to this list since we merged the clang-format thing was a checkbox to run clang-format on the code. There is (hopefully) about to be another PR that does a CI check for clang-format, but it's not in place yet, and not something I'm working on.

Author: In our conversation, we thought that because of https://github.com/cyclus/cyclus/pull/1674, we can pass on this

> - [ ] Address all review comments

Reviewer 4: What if we phrase these in a way such that the contributor is declaring that they have done these things? That way it's clear who needs to check these off. 

I also think the final check about review comments is something that reviewers will state when they're going through review. The checkbox might apply to multiple people so I think this might add unnecessary ambiguity. 
```suggestion
- [ ] I have read the [Contributing to Cyclus](https://fuelcycle.org/kernel/contributing_to_cyclus.html) guide
- [ ] I have compiled and run the code locally
- [ ] I have added or updated relevant tests
- [ ] I have added documentation for new or changed features
- [ ] This code follows the style guide
- [ ] I have updated the changelog
```

Reviewer 3: Change "add" on L32 to "added"
Change "update" on L35 to "updated"

otherwise I agree

Reviewer 4: Done!
